workflow/snakemake_rules/remote_files.smk

```diff
@@ -0,0 +1,89 @@
+"""Shared functions to transparently handle local and remote inputs/outputs.
+"""
+from functools import partial
+from importlib import import_module
+from urllib.parse import urlsplit, urljoin
+
+
+def _remote_producer(init):
+    """
+    Produce a remote input function by loading a :class:`snakemake.remote`
+    module and passing the module to an *init* function.
+
+    The module is named by the first argument of the *init* function.  *init*
+    will be called with the loaded module and is expected to return a remote
+    input function suitable for use inside :func:`path_or_url`.
+
+    Returns a "producer" function which when called with no arguments returns
+    the remote input function.
+
+    The call to *init* (that is, the production of the remote input function)
+    is deferred until the first time the producer is called.
+
+    If the remote module is not available because required dependencies aren't
+    installed, then a :exc:`WorkflowError` exception object is returned on the
+    first call of the producer.
+    """
+    # Name of first argument to *init*.
+    remote_name = init.__code__.co_varnames[0]
+    remote = None
+
+    def producer():
+        nonlocal remote
+        if not remote:
+            remote = init(import_module(f"snakemake.remote.{remote_name}"))
+        return remote
+
+    return producer
+
+
+SCHEME_REMOTES = {
+    "https": _remote_producer(lambda HTTP: HTTP.RemoteProvider().remote),
+    "http": _remote_producer(lambda HTTP: partial(HTTP.RemoteProvider().remote, insecure = True)),
+    "s3": _remote_producer(lambda S3: S3.RemoteProvider().remote),
+    "gs": _remote_producer(lambda GS: GS.RemoteProvider().remote),
+    "": lambda: lambda local_path, **kwargs: local_path,
+}
```

Comment:
I am curious whether there is a way in Snakemake to autodetect the provider from the scheme automagically and to not hardcode a subset (somewhat arbitrarily chosen. Why not Azure or Dropbox for example?), but to use everything available in Snakemake. Deferring the choice to Snakemake itself might also simplify the scheme/path logic below.

I don't know Snakemake very well, but I am surprised his is not automatic.
Could not find anything about autodetection in the docs either:
https://snakemake.readthedocs.io/en/stable/snakefiles/remote_files.html

Reply:
@ivan-aksamentov Snakemake tries to offer auto-detection through its [`AUTO` provider](https://snakemake.readthedocs.io/en/stable/snakefiles/remote_files.html#auto), but it's broken.

It enumerates and loads all the various `snakemake.remote.*` modules on each call to `AUTO.remote()`. This slows things down unnecessarily, but isn't a deal breaker. However, during enumeration, it protects against unsupported modules due to missing deps, but not supported modules which fail to initialize because of missing required credentials/config from the environment or missing required parameters to the remote provider or `AUTO.remote()`. This alone makes it completely unusable. I can't imagine how anyone is using it in this state. I suspect no one is.

(And even with a functioning `AUTO`, we'd still need some handling for local paths.)

Reply:
I'll add: We could make a better functioning `AUTO` equivalent ourselves, by similarly dynamically building the scheme → remote function mapping, but it didn't seem worth it to support use cases that aren't concrete right now. The implementation in this PR could certainly be further developed in that direction in the future, however.

Reply:
I added this reasoning as a comment at the top of `remote_files.smk` so it's recorded for future reconsideration. Thanks for spurring me to explain it.